From fce11f68491b46b93df69de0630cd9edb90bc772 Mon Sep 17 00:00:00 2001
From: INAGAKI Hiroshi <musashino.open@gmail.com>
Date: Thu Sep 29 02:13:04 2022 +0900
Subject: realtek: add SMP ops registration support to MIPS generic in 5.15

--- a/arch/mips/include/asm/machine.h
+++ b/arch/mips/include/asm/machine.h
@@ -16,6 +16,7 @@ struct mips_machine {
 	bool (*detect)(void);
 	const void *(*fixup_fdt)(const void *fdt, const void *match_data);
 	unsigned int (*measure_hpt_freq)(void);
+	void (*register_smp_ops)(void);
 };
 
 extern long __mips_machines_start;
--- a/arch/mips/generic/init.c
+++ b/arch/mips/generic/init.c
@@ -115,6 +115,12 @@ void __init device_tree_init(void)
 	unflatten_and_copy_device_tree();
 	mips_cpc_probe();
 
+	if (mach && mach->register_smp_ops) {
+		pr_info("Calling board specific registration of SMP ops\n");
+		mach->register_smp_ops();
+		return;
+	}
+
 	err = register_cps_smp_ops();
 	if (err)
 		err = register_up_smp_ops();
